lib/net/ip.go

```diff
@@ -115,8 +115,23 @@ func IPToBigInt(ip IP) *big.Int {
 }
 
 func BigIntToIP(ipInt *big.Int) IP {
-	ip := IP{net.IP(ipInt.Bytes())}
-	return ip
+	a := ipInt.Bytes()
+	ip := net.IP(a)
+	if ip4 := ip.To4(); ip4 != nil {
+		return IP{ip}
+	}
+
+	// for ipv6 IP like aa::1 and b::1 , the ipInt.Bytes() length of bigInt is 15 , shortter than 16,
+	// it will generate an invalid <nil> IP. So add the prefix 0 to make it succeed to convert to IP struct
+	if len(a) < 16 {
```

Comment:
I believe that the math/big.Int struct includes a function that will handle this for us.

From [the documentation for Bytes](https://golang.org/pkg/math/big/#Int.Bytes):

> To use a fixed length slice, or a preallocated one, use FillBytes.

So, I think we should do something like the following:

```
	// Use a fixed-length of 16 bytes for IPv6 addresses, since some IPv6 addresses can resolve to fewer 
	// than 16 bytes total, which is not properly interpreted as a net.IP. 
	a = make([]byte, 16)
	ipInt.FillBytes(a)
```

Reply:
the calico/go-build image seems to use go with version 1.14.4 , which still misses the big.Int FillBytes method . FillBytes is available on latest version like 1.15+

Reply:
Hm, ok. In that case, I'll try to get this updated for you: https://github.com/projectcalico/go-build/pull/137

Reply:
@weizhouBlue I've merged https://github.com/projectcalico/libcalico-go/pull/1316 which bumps the version of go to v1.15.2. If you rebase on master you should be able to pick it up!

Reply:
many thanks for the PR , and I see it
After rebase, however, found calico/go-build:v0.48 still uses go v1.14.4 ?   and the UT test still failed. Maybe need to rebuild a new verson of  calico/go-build  

# docker run  -it calico/go-build:v0.48 go version
Starting with UID : 9001
go version go1.14.4 linux/amd64


```
# make ut
.....
mkdir -p .go-pkg-cache bin /root/go/pkg/mod && docker run --rm --net=host --init -e RUN_AS_ROOT='true' -e GO111MODULE=on -v /root/go/pkg/mod:/go/pkg/mod:rw -e LOCAL_USER_ID=0 -e GOCACHE=/go-cache -e GOARCH=amd64 -e GOPATH=/go -e OS=linux -e GOOS=linux -e GOFLAGS="-mod=vendor" -v /root/libcalico-go:/go/src/github.com/projectcalico/libcalico-go:rw -v /root/libcalico-go/.go-pkg-cache:/go-cache:rw -w /go/src/github.com/projectcalico/libcalico-go --privileged calico/go-build:v0.48 \
	go version
**go version go1.14.4 linux/amd64**
mkdir -p .go-pkg-cache bin /root/go/pkg/mod && docker run --rm --net=host --init -e RUN_AS_ROOT='true' -e GO111MODULE=on -v /root/go/pkg/mod:/go/pkg/mod:rw -e LOCAL_USER_ID=0 -e GOCACHE=/go-cache -e GOARCH=amd64 -e GOPATH=/go -e OS=linux -e GOOS=linux -e GOFLAGS="-mod=vendor" -v /root/libcalico-go:/go/src/github.com/projectcalico/libcalico-go:rw -v /root/libcalico-go/.go-pkg-cache:/go-cache:rw -w /go/src/github.com/projectcalico/libcalico-go --privileged calico/go-build:v0.48 \
	sh -c 'cd /go/src/github.com/projectcalico/libcalico-go && ginkgo -r --skipPackage vendor -skip "\[Datastore\]" -focus=".*" .'
Failed to compile v3:

# github.com/projectcalico/libcalico-go/lib/net
**lib/net/ip.go:119:11: ipInt.FillBytes undefined (type *big.Int has no field or method FillBytes)**
```

Reply:
@weizhouBlue yes, that was my bad. I pushed the wrong code! 

It should be fixed now if you'd like to give it another shot. You'll need to clear out the old v0.48 image so that the new one can be pulled.

Reply:
```
$ docker run -ti calico/go-build:v0.48 go version                                                                                                                                                                                                    
Starting with UID : 9001
go version go1.15.2 linux/amd64
```